### ISSUE TITLE
Array Product Not Defined... How does Box.JS Get there?

### ISSUE THREAD

**Issue by Theory5** [CapacitorSet/box-js#13]
Hello! We've spoken before via email.

You pointed me to node.js after I answered a thread on reddit wanting to build another program. I wasn't experienced enough to help you though.

I know you've already spoken about how to get something to be defined in your readme, but I am confused. I do not know if it's simply because I'm inexperienced or what. 

Basically both firebug and node.js have one issue in common: they can puzzle out information from recursive arrays' but they do NOT show how they arrived at it!

So this malware I have is pretty much one giant array, with two smaller arrays. And somehow, it builds an entire program and config file out of them! I cannot seem to, no matter how much I try, figure this out!

Your program (and firebug) both output this function: _87867t67t6gt (but say it isn't defined) which is great, but I cannot seem to figure out HOW that function was generated from a set of random array stuff. 

What I was hoping for, was that your program would allow me to build the entire program from the array, THEN I could attempt to run it and see what it does via your software.

Is this a total misunderstanding of what the software does? if not, how can I see what it took from each array (and the calculations) to build the function and the rest of the program? Can I have it build EVERYTHING without attempting to run it (and thus stop it from running into the _____ not defined issue?).

Does this make any sense? Feel free to drop me a line at theoryofscience @ gmail . com, which we've talked through before!

**Comment by CapacitorSet** [comment]
Box-js is basically a wrapper around the V8 engine with some ActiveX emulators. Generally speaking, there is no such thing as "build the entire program but don't run it" - the program is built *by* running the script.

Now, in some cases you can actually get the program before it is executed; that's the case with droppers that build a new script first, and then evaluate. They usually look like this:

```js
var data = [...];
function process(input) { /* ... */ }
var script = process(data);
eval(script);
```

In these cases, it's a matter of finding the `eval` call (or more rarely, `new Function(script)()`), replacing it with `console.log` (so that the script will be printed to the output rather than evaluated), and then re-running the code.

Box-js can certainly be used for studying how a dropper was implemented, or debugging a dropper that isn't emulated correctly. It's usually a matter of using stack traces to see where an error occurs, messing around with variables and Proxies, and adding debug statements (`console.log(variable)`) in the target file. Depending on what exactly you're doing, it requires a certain level of knowledge of JavaScript.

---

In short: try to find the `eval(...)` call in your script and replace it with `console.log(...)`. Then, re-run the analysis, copy the output code, paste it in a new file, and analyze that. It helps a lot to format the code for easier readability; my goto tool is [the Esprima rewrite demo](http://esprima.org/demo/rewrite.html), which incidentally uses the same parsing engine as box-js.

If you need help with this, I could also take a look at the code directly, just send me the code either here or by mail.

**Comment by Theory5** [comment]
Awesome! It was user error I guess, thanks for the help. Changing the eval statement did... something, and it outputted a LOT of different things. The statements/functions in question are interesting to say the least. Apparently the function I posted above is a variable for "Environment system" stuff.

So you can follow along at home, here is a link to the malware code (in pastebin, not a js file):

http://pastebin.com/jpYTFYVE

And the output in question:

http://pastebin.com/FM4EpAjf

Cool stuff, huh?

Thanks again!

**Comment by CapacitorSet** [comment]
Nice to see it works!

By the way, the output is a very common category of JScript droppers, I've been seeing it since a few months. Here are a few pointers from a walkthrough I wrote some time ago; unfortunately, the line numbers are different, and I was explicitly asked not to share the sample.

>First of all, you'll notice that the "actual code" is wrapped in an eval call (line 11). If you replace that with a console.log, print the output to a file, and beautify, you get the actual code. The first part (L1..13) checks that the environment variable ComSpec equals cmd.exe; if so, it proceeds with the download: L332..344 set the download method to either XMLHTTP or WScriptNetwork, whichever is available (note that both expose the same APIs). The for loop at L350 tries each of the URLs. L361..365 download the file. Now, note that Uv defined in L366 is the ActiveX object "ADODB.Stream". L369 writes the result of the download to Uv, and L371 saves it to a file with path Lp9.

>Now, the actual decoding takes part. Remember that at this point, the encoded file was saves to Lp9. L373 calls HIi(Lp9), which (L412..422) creates a new ADODB.Stream, reads from the path Lp9, and fetches its text. The codepage 437 comes into play here: L415 sets Charset to 437. The text then goes through the function IGi2 (L423..566), which contains a map from 1- or 2-byte values to 1-byte values (L425..552). For each character of the text (L554..563), if it is smaller than 0x007F it produces the same byte, else it looks it up in the map and produces the resulting value. The result of this decoding is stored to CJf2 at L373. Then, Kx1 (L392..411) does the decryption: it XORs each byte with a pseudorandom stream with a fixed seed TIk (L395..397). Then, it performs a simple checksum. First of all, it takes the last four bytes (L398-9) out of the string, which is the "expected checksum". It computes the "actual checksum", which is an initial value UFn4, summed with every byte, modulo 1000000000 (L401..4). The actual checksum is expected to equal the expected checksum, and if it does, the decrypted text is returned. L375 performs one last check: the size of the decrypted text is expected to be within certain limits.

>If you wanted to write a decoder for yourself, you could probably reuse the JavaScript code, which can be transpiled very easily to other languages. The only issue is implementing codepage-437 support, which is something I struggled with because of a relative lack of documentation, and determining the seed for the PRNG. If you read the source code for ADODBStream.js, you'll see that it contains code from HynekPetrak/malware-jail (which is also a good sandbox, btw), which had already implemented it correctly. In particular, here is the commit where I added emulation for codepage-437: it uses iconv to encode the stream when it is written to.
